src/Orleans.Runtime/Placement/ResourceOptimizedPlacementDirector.cs

```diff
@@ -0,0 +1,339 @@
+using System;
+using System.Buffers;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Options;
+using Orleans.Runtime.Configuration.Options;
+
+namespace Orleans.Runtime.Placement;
+
+// See: https://www.ledjonbehluli.com/posts/orleans_resource_placement_kalman/
+internal sealed class ResourceOptimizedPlacementDirector : IPlacementDirector, ISiloStatisticsChangeListener
+{
+    /// <summary>
+    /// 1 / (1024 * 1024)
+    /// </summary>
+    private const float PhysicalMemoryScalingFactor = 0.00000095367431640625f;
+    private const int FourKiloByte = 4096;
+
+    private readonly NormalizedWeights _weights;
+    private readonly float _localSiloPreferenceMargin;
+    private readonly ConcurrentDictionary<SiloAddress, FilteredSiloStatistics> _siloStatistics = [];
+
+    private Task<SiloAddress> _cachedLocalSilo;
+
+    public ResourceOptimizedPlacementDirector(
+        DeploymentLoadPublisher deploymentLoadPublisher,
+        IOptions<ResourceOptimizedPlacementOptions> options)
+    {
+        _weights = NormalizeWeights(options.Value);
+        _localSiloPreferenceMargin = options.Value.LocalSiloPreferenceMargin;
+        deploymentLoadPublisher.SubscribeToStatisticsChangeEvents(this);
+    }
+
+    private static NormalizedWeights NormalizeWeights(ResourceOptimizedPlacementOptions input)
+    {
+        var totalWeight = input.CpuUsageWeight + input.MemoryUsageWeight + input.PhysicalMemoryWeight + input.AvailableMemoryWeight;
```

Comment:
Added fix in latest commit to avoid a `DivideByZeroException` which would be possible in case the sum of all weights be 0